### ISSUE TITLE
EISDIR illegal operation on a dirrectory, read errno -21

### ISSUE THREAD

**Issue by zirmax** [franciscop/legally#24]
Hello,

i tried to run legally (v3.3.2) on following packages:

```
  "dependencies": {
    "axios": "0.18.0",
    "node-port-check": "2.0.0",
    "redis": "2.8.0",
    "socket.io-client": "2.1.1",
    "winston": "3.0.0"
  },
  "devDependencies": {
    "@babel/plugin-transform-destructuring": "7.0.0",
    "babel-cli": "6.26.0",
    "babel-plugin-transform-object-rest-spread": "6.26.0",
    "babel-polyfill": "6.26.0",
    "babel-preset-env": "1.7.0",
    "babel-register": "6.26.0",
    "chai": "4.1.2",
    "esdoc": "1.1.0",
    "esdoc-standard-plugin": "1.0.0",
    "eslint": "5.3.0",
    "eslint-config-standard": "11.0.0",
    "eslint-plugin-import": "2.14.0",
    "eslint-plugin-mocha": "5.2.0",
    "eslint-plugin-node": "7.0.1",
    "eslint-plugin-promise": "3.8.0",
    "eslint-plugin-standard": "3.1.0",
    "mocha": "5.2.0",
    "moxios": "0.4.0",
    "nyc": "12.0.2",
    "sinon": "6.1.5",
    "standard": "11.0.1"
  },
```

But i get the following Error:
```
Working on ".". It might take a while...
{ Error: EISDIR: illegal operation on a directory, read errno: -21, code: 'EISDIR', syscall: 'read' } { _: [], routes: [] }
(node:812) UnhandledPromiseRejectionWarning: Error: EISDIR: illegal operation on a directory, read
(node:812) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:812) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by michaelrambeau** [comment]
Hello @franciscop ,
I had the same problem when I checked a package called `baconjs`, that has no dependency.

```js
const legally = require("legally")
await legally('baconjs')
```

```
EISDIR: illegal operation on a directory, read
```

**Comment by michaelrambeau** [comment]
Hello @franciscop 
The result of my investigation: the error is caused by the folder called `readme`, there is an attempt to open it as a file (it confirms my previous comment above).

I did some tests in local and it seems that the problem could be solved by using `walk` (that excludes directories) instead of `list` in this line: https://github.com/franciscop/legally/blob/master/src/legally.js#L6

```js
const search = (root, regex) => list(root) // should be `walk(root)` instead?
  .filter(file => regex.test(file))
  .map(read)
  .map(searchText)
  .reduce((all, one) => [...all, ...one], []);
```

But I'm not 100% sure about the algorithm, let me know if you think it's OK.
I can create a PR if you want.

**Comment by franciscop** [comment]
Should be working now with `legally@3.3.3`. I added this fix:

```js
  .filter(file => stat(file).isFile())
```

I had to modify `magic-promises` to [allow for async filter](https://github.com/franciscop/magic-promises/issues/1), then upgraded `fs-array` to include the right dependency and finally it came all the way here to be able to add that filter. Sorry it took a while, I've been quite busy this weekend.

**Comment by zirmax** [comment]
Works now, big thanks :)

**Comment by michaelrambeau** [comment]
Thank you Francisco, `await legally("baconjs")` works (tested in v3.3.3), good job!

Just for the records, as we discussed, I've seen that now you first check that the item is a file (and not a folder) and then you check its name using the RegExp.
I suppose that the first check is far much expensive than the second one, so it could be faster to do the opposite: first check the item name and then check if it's a file and not a folder.

**Comment by franciscop** [comment]
Just measured, the difference seems to be trivial for `express` (already installed, just reading):
- Stat first: 0.541 +- 0.011
- Stat later: 0.577 +- 0.008
- Diff: 6.7%

Since it doesn't hurt, I've published it on `3.3.4`. But no need to upgrade for this, it's negligible.

**Comment by michaelrambeau** [comment]
Thank you for the benchmark!
